File: csharp/1130.minimum-cost-tree-from-leaf-values.cs
/*
 * @lc app=leetcode id=1130 lang=csharp
 *
 * [1130] Minimum Cost Tree From Leaf Values
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int MctFromLeafValues(int[] arr) {
        var stack = new Stack<int>();
        int ret = 0;
        foreach (var num in arr) {
            while (stack.TryPeek(out int top) && top <= num) {
                stack.Pop();
                if (stack.TryPeek(out int next) && next <= num) {
                    ret += top * next;
                } else {
                    ret += top * num;
                    break;
                }
            }
            stack.Push(num);
        }
        while (stack.Count >= 2) {
            ret += stack.Pop() * stack.Peek();
        }
        return ret;
    }
}
// @lc code=end

File: java/1130.minimum-cost-tree-from-leaf-values.java
/*
 * @lc app=leetcode id=1130 lang=java
 *
 * [1130] Minimum Cost Tree From Leaf Values
 */

import java.util.Stack;

// @lc code=start
class Solution {
    public int mctFromLeafValues(int[] arr) {
        Stack<Integer> stack = new Stack<>();
        int ret = 0;
        for (int num : arr) {
            while (!stack.isEmpty() && stack.peek() <= num) {
                int pop = stack.pop();
                if (!stack.isEmpty() && stack.peek() <= num) {
                    ret += pop * stack.peek();
                } else {
                    ret += pop * num;
                    break;
                }
            }
            stack.push(num);
        }
        while (stack.size() > 1) {
            ret += stack.pop() * stack.peek();
        }
        return ret;
    }
}
// @lc code=end

